### PR TITLE
Add IM protocol code

### PR DESCRIPTION
**Pull request by yunhanw-google**
#### Problem
Currently the IM protocol code has not yet been implemented

#### Change overview
Add IM protocol code per IM spec

#### Testing
the existing test cover this.

### CODE REVIEW

**Review by andy31415 — APPROVED** [review]

**Comment by bzbarsky-apple** [comment]
> Add IM protocol code per IM spec

@yunhanw-google which specific part(s) of the spec is this aiming to implement?

**Comment by yunhanw-google** [comment]
> > Add IM protocol code per IM spec
> 
> @yunhanw-google which specific part(s) of the spec is this aiming to implement?

https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/a9781609d101256455fd7ce52ec2b28e5853a012/src/data_model/Interaction-Model.adoc#status-code-table
This table record status code particularly for IM

File: src/protocols/interaction_model/Constants.h [hunk 1]
```diff
@@ -62,6 +62,32 @@ enum class MsgType : uint8_t
...
+    kInvalidSubscription    = 0x7d,
+    kUnsupportedAccess      = 0x7e,
+    kUnsupportedEndpoint    = 0x7f,
+    kInvalidAction          = 0x80,
+    kUnsupportedCommand     = 0x81,
+    kInvalidCommand         = 0x85,
+    kUnsupportedAttribute   = 0x86,
+    kInvalidValue           = 0x87,
+    kUnsupportedWrite       = 0x88,
+    kResourceExhansted      = 0x89,
```
Comment (woody-apple) [inline src/protocols/interaction_model/Constants.h:78]: Typo :)

File: src/protocols/interaction_model/Constants.h [hunk 2]
```diff
@@ -62,6 +62,32 @@ enum class MsgType : uint8_t
     TimedRequest          = 0x0a,
 };
 
+enum class ProtocolCode : uint16_t
```
Comment (woody-apple) [inline src/protocols/interaction_model/Constants.h:65]: Where are these defined from?

File: src/app/util/ember-compatibility-functions.cpp [hunk 3]
```diff
@@ -71,11 +72,15 @@ bool IMEmberAfSendDefaultResponseWithCallback(EmberAfStatus status)
                                                        imCompatibilityEmberAfCluster.commandId,
                                                        (chip::app::CommandPathFlags::kEndpointIdValid) };
 
+    if (status != EMBER_ZCL_STATUS_SUCCESS)
```
Comment (woody-apple) [inline src/app/util/ember-compatibility-functions.cpp:75]: Are we losing clarity on what time of error this is? Looks like we're dropping resolution of the error, without even logging?

**Review by woody-apple — CHANGES_REQUESTED** [review]
Looks like there was an error case added here, that existing tests don't cover. Can we also add a test there?

File: src/protocols/interaction_model/Constants.h (hunk 2 above)
Comment (yunhanw-google) [reply src/protocols/interaction_model/Constants.h:65]: they are defined in https://github.com/CHIP-Specifications/connectedhomeip-spec/blob/a9781609d101256455fd7ce52ec2b28e5853a012/src/data_model/Interaction-Model.adoc#status-code-table

File: src/app/util/ember-compatibility-functions.cpp (hunk 3 above)
Comment (yunhanw-google) [reply src/app/util/ember-compatibility-functions.cpp:75]: currently ember code is not 1on1 matching with IM status code. maybe i can modify the ember code to align with IM ones.

File: src/protocols/interaction_model/Constants.h (hunk 1 above)
Comment (yunhanw-google) [reply src/protocols/interaction_model/Constants.h:78]: thanks for catching this issue。

Comment (yunhanw-google) [reply src/protocols/interaction_model/Constants.h]: Done, thanks

File: src/protocols/interaction_model/Constants.h [hunk 4]
```diff
@@ -62,6 +62,49 @@ enum class MsgType : uint8_t
...
+    kSuccess                = 0x0,
+    kFailure                = 0x01,
+    kInvalidSubscription    = 0x7d,
+    kUnsupportedAccess      = 0x7e,
+    kUnsupportedEndpoint    = 0x7f,
+    kInvalidAction          = 0x80,
+    kUnsupportedCommand     = 0x81,
+    kReservedUnsuptedGeneralCommand   = 0x82,
+    kReservedUnsupManufClusterCommand = 0x83,
+    kReservedUnsupManufGeneralCommand = 0x84,
```
Comment (bzbarsky-apple) [inline src/protocols/interaction_model/Constants.h:77]: If we can avoid assigning any semantically meaningful names to the reserved values, that would be ideal.  So `Reserved1`, `Reserved2`, etc.  Or if desired `Reserved82 = 0x82` and so forth. Or just not add them to the enum at all.

File: src/protocols/interaction_model/Constants.h [hunk 5]
```diff
@@ -62,6 +62,49 @@ enum class MsgType : uint8_t
...
+    kInvalidCommand         = 0x85,
+    kUnsupportedAttribute   = 0x86,
+    kInvalidValue           = 0x87,
+    kUnsupportedWrite       = 0x88,
+    kResourceExhausted      = 0x89,
+    kReservedDuplicateExists = 0x8a,
+    kNotFound               = 0x8b,
+    kUnreportableAttribute  = 0x8c,
+    kInvalidDataType        = 0x8d,
+    kUnsupportedRead        = 0x8f,
```
Comment (bzbarsky-apple) [inline src/protocols/interaction_model/Constants.h]: The reserved `8e` is missing.  We should either include it or remove the other reserved values.

File: src/protocols/interaction_model/Constants.h [hunk 6]
```diff
@@ -62,6 +62,49 @@ enum class MsgType : uint8_t
...
+    kReservedInconsistentStartUpState = 0x90,
+    kReservedOutOfBand        = 0x91,
+    kReservedActionDenied              = 0x93,
```
Comment (bzbarsky-apple) [inline src/protocols/interaction_model/Constants.h]: Reserved `0x92` is missing.

File: src/protocols/interaction_model/Constants.h [hunk 7]
```diff
@@ -62,6 +62,49 @@ enum class MsgType : uint8_t
...
+    kTimeout                = 0x94,
+    kReservedAbort                       = 0x95,
+    kReservedInvalidImage             = 0x96,
+    kReservedWaitForData               = 0x97,
+    kReservedNoImageAvailable          = 0x98,
+    kReservedRequiredMoreImage          = 0x99,
+    kReservedNotificationPending        = 0x9A,
```
Comment (bzbarsky-apple) [inline src/protocols/interaction_model/Constants.h]: Can we be consistent about casing?  The others are lowercase...

File: src/protocols/interaction_model/Constants.h [hunk 8]
```diff
@@ -62,6 +62,49 @@ enum class MsgType : uint8_t
...
+    kConstraintError        = 0x9b,
+    kBusy                   = 0x9c,
+    kReservedHardwareFailure            = 0xC0,
+    kReservedSoftwareFailure            = 0xC1,
```
Comment (bzbarsky-apple) [inline src/protocols/interaction_model/Constants.h]: Again, consistent case, please.

File: src/protocols/interaction_model/Constants.h [hunk 9]
```diff
@@ -62,6 +62,49 @@ enum class MsgType : uint8_t
...
+    kUnsupportedCluster     = 0xc3,
+    kReservedUnSupportedCluster         = 0xC3,
+    kNoUpstreamSubscription = 0xc5,
```
Comment (bzbarsky-apple) [inline src/protocols/interaction_model/Constants.h]: `0xc4` missing.

File: src/protocols/interaction_model/Constants.h [hunk 10]
```diff
@@ -62,6 +62,49 @@ enum class MsgType : uint8_t
     TimedRequest          = 0x0a,
 };
 
+//TODO: Cleanup Reserved protocolCode from ember
+enum class ProtocolCode : uint16_t
```
Comment (bzbarsky-apple) [inline src/protocols/interaction_model/Constants.h:66]: Please document that this table comes from the IM's "Status Code Table" section from the spec.

**Review by bzbarsky-apple — CHANGES_REQUESTED** [review]

File: src/app/CommandHandler.cpp [hunk 11]
```diff
@@ -123,7 +123,7 @@ CHIP_ERROR CommandHandler::ProcessCommandDataElement(CommandDataElement::Parser
 
         AddStatusCode(&returnStatusParam,
                       err == CHIP_ERROR_INVALID_PROFILE_ID ? GeneralStatusCode::kNotFound : GeneralStatusCode::kInvalidArgument,
-                      Protocols::SecureChannel::Id, Protocols::SecureChannel::kProtocolCodeGeneralFailure);
+                      Protocols::InteractionModel::Id, Protocols::InteractionModel::ProtocolCode::kInvalidArgument);
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/CommandHandler.cpp]: Follow up on this https://github.com/project-chip/connectedhomeip/issues/7612

File: src/app/CommandHandler.cpp [hunk 12]
```diff
@@ -142,7 +142,7 @@ CHIP_ERROR CommandHandler::AddStatusCode(const CommandPathParams * apCommandPath
 
     statusElementBuilder =
         mInvokeCommandBuilder.GetCommandListBuilder().GetCommandDataElementBuilder().CreateStatusElementBuilder();
-    statusElementBuilder.EncodeStatusElement(aGeneralCode, aProtocolId.ToFullyQualifiedSpecForm(), aProtocolCode)
+    statusElementBuilder.EncodeStatusElement(aGeneralCode, aProtocolId.ToFullyQualifiedSpecForm(), static_cast<uint16_t>(aProtocolCode))
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/CommandHandler.cpp]: Done, thanks

File: src/protocols/interaction_model/Constants.h [hunk 13]
```diff
@@ -62,6 +62,49 @@ enum class MsgType : uint8_t
...
+{
+    kSuccess                = 0x0,
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/protocols/interaction_model/Constants.h]: Done, thanks

File: src/protocols/interaction_model/Constants.h (hunk 4 above)
Comment (yunhanw-google) [reply src/protocols/interaction_model/Constants.h]: Done, thanks

File: src/protocols/interaction_model/Constants.h (hunk 5 above)
Comment (yunhanw-google) [reply src/protocols/interaction_model/Constants.h]: include it , thanks

File: src/protocols/interaction_model/Constants.h (hunk 6 above)
Comment (yunhanw-google) [reply src/protocols/interaction_model/Constants.h]: include it , thanks

File: src/protocols/interaction_model/Constants.h (hunk 7 above)
Comment (yunhanw-google) [reply src/protocols/interaction_model/Constants.h]: Done, thanks

File: src/protocols/interaction_model/Constants.h (hunk 8 above)
Comment (yunhanw-google) [reply src/protocols/interaction_model/Constants.h]: Done, thanks

File: src/protocols/interaction_model/Constants.h (hunk 9 above)
Comment (yunhanw-google) [reply src/protocols/interaction_model/Constants.h]: Done, thanks

File: src/protocols/interaction_model/Constants.h (hunk 10 above)
Comment (yunhanw-google) [reply src/protocols/interaction_model/Constants.h:66]: Done, thanks

File: src/app/zap-templates/templates/app/im-cluster-command-handler.zapt [hunk 14]
```diff
@@ -62,8 +62,8 @@ void Dispatch{{asCamelCased side false}}Command(app::Command * apCommandObj, Com
                                                        ZCL_{{asDelimitedMacro define}}_ID,
                                                        aCommandId,
                                                        (chip::app::CommandPathFlags::kEndpointIdValid) };
-            apCommandObj->AddStatusCode(&returnStatusParam, Protocols::SecureChannel::GeneralStatusCode::kNotFound, Protocols::SecureChannel::Id,
-                            Protocols::SecureChannel::kProtocolCodeGeneralFailure);
+            apCommandObj->AddStatusCode(&returnStatusParam, Protocols::SecureChannel::GeneralStatusCode::kNotFound, Protocols::InteractionModel::Id,
+        Protocols::InteractionModel::ProtocolCode::kInvalidCommand);
```
[the remark this answers is not in the document]
Comment (yunhanw-google) [reply src/app/zap-templates/templates/app/im-cluster-command-handler.zapt]: Done, thanks

**Review by bzbarsky-apple — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

**Review by msandstedt — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

**Pull request merged**